zuchini-examples/zuchini-examples-mockmvc/src/test/java/org/zuchini/examples/mockmvc/MockMvcSteps.java

```diff
@@ -0,0 +1,102 @@
+
+    private HttpHeaders requestHeaders = new HttpHeaders();
+    private ResultActions resultActions;
+
+    private static MultiValueMap<String, String> toMultiValueMap(Datatable datatable) {
+        LinkedMultiValueMap<String, String> map = new LinkedMultiValueMap<>();
+        for (List<String> header : datatable.getRows()) {
+            Iterator<String> iterator = header.iterator();
+            if (!iterator.hasNext()) {
+                throw new IllegalStateException("Name is required in datatable");
+            }
+
+            String headerName = iterator.next();
+            while (iterator.hasNext()) {
+                String headerValue = iterator.next();
+                if (!headerValue.isEmpty()) {
+                    map.add(headerName, headerValue);
+                }
+            }
+        }
+        return map;
+    }
+
+    @Given("^the following request headers:$")
+    public void givenRequestHeaders(Datatable datatable) {
+        requestHeaders.putAll(toMultiValueMap(datatable));
+    }
+
+    @When("^a (GET|POST|PUT|DELETE) request to \"([^\"]+)\" is executed$")
+    public void whenRequestIsExecuted(String method, String uri) throws Exception {
+        resultActions = mvc.perform(request(HttpMethod.valueOf(method), URI.create(uri))
+                .headers(requestHeaders));
+    }
+
+    @When("^a (GET|POST|PUT|DELETE) request to \"([^\"]+)\" is executed with the following parameters:$")
+    public void whenRequestIsExecuted(String method, String uri, Datatable datatable) throws Exception {
+        resultActions = mvc.perform(request(HttpMethod.valueOf(method), URI.create(uri))
+                .headers(requestHeaders)
+                .params(toMultiValueMap(datatable)));
+    }
+
+    @Then("^the response status is ([0-9]+)$")
+    public void responseStatusIs(int status) throws Exception {
+        resultActions.andExpect(status().is(status));
+    }
+
+    @Then("^the response body matches the following json paths:$")
+    public void responseMatchesJsonPath(Datatable datatable) throws Exception {
+        MultiValueMap<String, String> map = toMultiValueMap(datatable);
+        for (Map.Entry<String, List<String>> entry : map.entrySet()) {
+            String key = entry.getKey();
+            JsonPathResultMatchers jsonPath = jsonPath(key);
+            for (String value : entry.getValue()) {
+                resultActions.andExpect(jsonPath.value(value));
+            }
+        }
+    }
+
+    @Then("^the response contains the following headers:$")
+    public void responseHeadersContain(Datatable datatable) throws Exception {
+        MultiValueMap<String, String> map = toMultiValueMap(datatable);
+        HeaderResultMatchers header = header();
+        for (Map.Entry<String, List<String>> entry : map.entrySet()) {
+            String key = entry.getKey();
+            for (String value : entry.getValue()) {
+                resultActions.andExpect(header.string(key, value));
```

Comment:
I always thought that `andExpect` is declarative, but it looks like it actually executes/verifies when invoked.